assembly/src/assembler/instruction/mod.rs

```diff
@@ -22,17 +25,18 @@ impl Assembler {
     pub(super) fn compile_instruction(
         &self,
         instruction: &Instruction,
-        span_builder: &mut SpanBuilder,
+        span_builder: &mut BasicBlockBuilder,
         ctx: &mut AssemblyContext,
-    ) -> Result<Option<CodeBlock>, AssemblyError> {
+        mast_forest: &mut MastForest,
```

Comment:
Question: should `mast_forest` be a part of the `AssemblyContext`? Seems like most of the time they are passed around together.

Reply:
I understood `AssemblyContext` to be more about metadata, and so put `MastForest` in `Assembler` instead. But I'd be curious to see what @bitwalker thinks.

Reply:
I am thinking about it as more like "anything that's needed during assembly of a specific program but doesn't need to be persisted after that." So, in my mind, MAST of the program being currently assembled would go into `AssemblyContext` - but also, we can come back to this in the next PR.

Reply:
Stuff that can be reused across compilations should go in `Assembler` IMO, so as to avoid recomputing a bunch of stuff we already know - but the degree to which that is used/useful at the moment is pretty limited, as we are typically instantiating/compiling/discarding the `Assembler` for each program individually in one go. The `AssemblyContext` is, as @plafer pointed out, mostly about configuration and what not for a single invocation of the `Assembler` APIs.

In my opinion, we should define how the `Assembler` is meant to be used (single-use vs multi-use compilation, the latter being useful when compiling many programs that use the same, or many of the same, dependencies). Once that decision is made, we can clean up the `Assembler` API and tailor it for how it is meant to be used. Right now it is having a bit of an identity crisis.

IMO, we should probably just make the `Assembler` single-use, and get rid of `AssemblyContext`. If we want to support the multiple-use case, we probably want to move more stuff into the `AssemblyContext`.